### PR TITLE
Issue 1153

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
for: https://github.com/openstax/unified/issues/1153

~~TODO:~~
- [x] Implement arrow background: https://openstax.slack.com/archives/C8V1XET1R/p1594041867447800
- [x] Check this on mobile

Mobile design as discussed here: https://openstax.slack.com/archives/C8V1XET1R/p1594394515046500

### CODE REVIEW

File: src/app/content/studyGuides/components/StudyGuidesCTA/styles.tsx [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+export const StudyGuidesCTAArrowDesktop = styled.img`
+  position: absolute;
+
+  ${theme.breakpoints.mobile(css`
+    display: none;
+  `)}
+`;
+
+// tslint:disable-next-line: variable-name
+export const StudyGuidesCTAArrowMobile = styled.img`
```
Comment (janKozie1) [inline src/app/content/studyGuides/components/StudyGuidesCTA/styles.tsx:150]: I think it should be possible to use them as background images instead and maybe it would be easier to center them over text. That said, i think this approach is fine too.

**Review by janKozie1 — COMMENTED** [review]
I think it looks good, just left minor comments

**Comment by janKozie1** [comment]
There are also some linting errors

File: src/app/content/studyGuides/components/StudyGuidesCTA/index.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAContent)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTATitle)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAButtons)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAButton)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTASeparator)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTALink)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAInfoWrapper)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAInfo)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAArrowDesktop)).not.toThrow();
+    expect(() => component.root.findByType(Styled.StudyGuidesCTAArrowMobile)).not.toThrow();
```
[the remark this answers is not in the document]
Comment (PiotrKozlowski) [reply src/app/content/studyGuides/components/StudyGuidesCTA/index.spec.tsx:49]: I think you've suggested on another pr to check for specific elements rather than using snapshots because it is hard to follow what's missing after snapshot update. I think checking for specific items is okay, but yea, it's a long list.

File: src/app/content/utils/archiveTreeUtils.spec.ts [hunk 3]
```diff
@@ -174,3 +175,25 @@ describe('nodeHasId', () => {
...
 });
+
+describe('comparePositionsOfNodes', () => {
+  it('return 0 for the same ids', () => {
+    expect(comparePositionsOfNodes(book.tree, '1', '1')).toEqual(0);
+  });
+
+  it('return 0 for ids that does not belongs to this book', () => {
+    expect(comparePositionsOfNodes(book.tree, '1', '2')).toEqual(0);
+  });
```
[the remark this answers is not in the document]
Comment (PiotrKozlowski) [reply src/app/content/utils/archiveTreeUtils.spec.ts:186]: I think this is more consistent with `if (idA === idB) { return 0; }`, because provided ids may also not belong to this book, but I don't have a strong opinion on this so I've added an error at the end of the function

File: src/app/content/studyGuides/components/StudyGuidesCTA/index.spec.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (janKozie1) [reply src/app/content/studyGuides/components/StudyGuidesCTA/index.spec.tsx:49]: I did, but i feel like it's not necessary to assert every component because they are not all separately conditionally rendered. But you are right too, if not every element is asserted then its still possible to miss it after updating. 

I'm not sure what's the better approach then. Snapshots are usually brittle and easy to miss, assertions on each component are tedious. What do you think?

File: src/app/content/studyGuides/components/StudyGuidesCTA/index.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  if (!isNotLoggedIn) { return null; }
+
+  return <Styled.StudyGuidesCTAWrapper>
+    <Styled.StudyGuidesCTAContent>
+      <StudyGuidesCTATitle />
+      <Styled.StudyGuidesCTAButtons>
+        <FormattedMessage id='i18n:studyguides:cta:button'>
+          {(msg: Element | string) => <Styled.StudyGuidesCTAButton
+            // eslint-disable-next-line
```
[the remark this answers is not in the document]
Comment (PiotrKozlowski) [reply src/app/content/studyGuides/components/StudyGuidesCTA/index.tsx]: PS: `Line 34:  Anchors must have content and the content must be accessible by a screen reader  jsx-a11y/anchor-has-content`

[the remark this answers is not in the document]
Comment (PiotrKozlowski) [reply src/app/content/studyGuides/components/StudyGuidesCTA/index.tsx]: That's from the `yarn build`

File: src/app/content/studyGuides/components/StudyGuidesCTA/index.spec.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (PiotrKozlowski) [reply src/app/content/studyGuides/components/StudyGuidesCTA/index.spec.tsx]: I think it doesn't really matter, but snapshots have one more advantage - they show you a text inside components, aria labels, etc. So I think we can go with a snapshot in this one, especially that it will probably not be changing anytime soon

File: src/app/content/utils/archiveTreeUtils.spec.ts [hunk 5]
```diff
@@ -174,3 +175,25 @@ describe('nodeHasId', () => {
     expect(nodeHasId('some-id', node)).toEqual(false);
   });
 });
+
+describe('comparePositionsOfNodes', () => {
+  it('return 0 for the same ids', () => {
+    expect(comparePositionsOfNodes(book.tree, '1', '1')).toEqual(0);
+  });
+
+  it('throws an error if non of provided ids does not belong to this book', () => {
```
Comment (janKozie1) [inline src/app/content/utils/archiveTreeUtils.spec.ts:184]: `none`

File: src/app/content/utils/archiveTreeUtils.ts [hunk 6]
```diff
@@ -123,3 +123,18 @@ export const archiveTreeSectionIsChapter = (section: LinkedArchiveTreeNode): sec
...
   && getArchiveTreeSectionNumber(section) !== null
 ;
+
+/**
+ * If @param idA is before @param idB then return -1
+ * If @param idB is before @param idA then return 1
+ * In any other case return 0
+ */
+export const comparePositionsOfNodes = (tree: LinkedArchiveTree, idA: string, idB: string) => {
+  if (idA === idB) { return 0; }
```
Comment (janKozie1) [inline src/app/content/utils/archiveTreeUtils.ts:133]: I feel like this can lead to some misleading results if the ids somehow happen to be in a different book. I don't have a strong opinion on it tho.

**Review by janKozie1 — APPROVED** [review]
I think it looks good.
Do you think something similar to https://github.com/openstax/unified/issues/1107#issuecomment-655653303 applies to this pr?

File: src/app/content/utils/archiveTreeUtils.ts (hunk 6 above)
Comment (PiotrKozlowski) [reply src/app/content/utils/archiveTreeUtils.ts:133]: I was thinking about this at I think it may be a good thing to not throw an error if ids are not found in the book, because we can't know if what we pass in the development is a valid child of a book so this can cause errors on the production (for example comparing ids returned from the api). I think returning 0 when ids were not found is okay, but we can wait for @TomWoodward to decide

**Comment by PiotrKozlowski** [comment]
@janKozie1 
> Do you think something similar to openstax/unified#1107 (comment) applies to this pr?

Maybe. @TomWoodward Should I apply the same styles for this pr?
https://github.com/openstax/unified/issues/1107#issuecomment-655746449

**Comment by TomWoodward** [comment]
> @janKozie1
> 
> > Do you think something similar to [openstax/unified#1107](https://github.com/openstax/unified/issues/1107) (comment) applies to this pr?
> 
> Maybe. @TomWoodward Should I apply the same styles for this pr?
> [openstax/unified#1107 (comment)](https://github.com/openstax/unified/issues/1107#issuecomment-655746449)

what would it apply to? you're not adding any popups right?

**Comment by PiotrKozlowski** [comment]
@TomWoodward Mobile version for this CTA looks kind of weird on larger screens
image

**Comment by TomWoodward** [comment]
> @TomWoodward Mobile version for this CTA looks kind of weird on larger screens
> image

in the designs it looks like the arrow is supposed to get bigger on desktop, which might help. isn't the modal itself fixed width? so we shouldn't be dealing with anything super super wide

**Comment by PiotrKozlowski** [comment]
@TomWoodward this screen is from mobile resolution, about 1190px

**Comment by TomWoodward** [comment]
> @TomWoodward this screen is from mobile resolution, about 1190px
> on a desktop it looks okay

ah... yea.. idk, we could make this CTA use a different breakpoint so that it kicks over to the bigger arrow and margins as soon as it looks good, or we could try to rewrite it so it scales proportionally. you may want to ping suhani to see if its worth looking into that or not

File: src/app/content/studyGuides/hooks/openStudyGuides.ts [hunk 7]
```diff
@@ -8,8 +9,13 @@ export const hookBody: ActionHookBody<typeof openStudyGuides> = (services) => as
   const filtersHaveBeenSet = select.filtersHaveBeenSet(state);
   const currentFilters = select.summaryLocationFilters(state);
   const defaultFilter = select.defaultLocationFilter(state);
+  const notLoggedIn = loggedOut(state);
+  const locationFilters = select.studyGuidesLocationFiltersWithContent(state);
+  const firstChapter = Array.from(locationFilters)[0];
```
Comment (TomWoodward) [inline src/app/content/studyGuides/hooks/openStudyGuides.ts:14]: `getHighlightLocationFilterForPage` would work if you passed a chapter into it, you could rename that, pull the first chapter out of some archive util and get the correct chapter one filter that way, rather than adding all the sorting logic to the filter utils, i think that might be cleaner

**Review by TomWoodward — CHANGES_REQUESTED** [review]

**Review by TomWoodward — APPROVED** [review]
